test/box-luatest/gh_6794_recover_nonmatching_xlogs_test.lua

```diff
@@ -4,9 +4,54 @@ local fio = require('fio')
 
 local g = t.group()
 
+g.before_all(function(cg)
+    -- We need to create such an xlog and a snap that the xlog has data both
+    -- before and after the snap. For example, 0...0.xlog and 0...02.snap, but
+    -- the xlog contains entries with lsns 1, 2 and 3. This is not possible with
+    -- one instance, because it rotates its xlog while making a snapshot. So,
+    -- create two instances with identical uuids, and make one of them generate
+    -- the needed snapshot and the other - the needed xlog. Then move the files
+    -- into one folder.
+    cg.box_cfg = {instance_uuid = '6f9019aa-2821-4f34-905b-51aed43def47'}
+    cg.snap = server:new{alias = 'snapshot', box_cfg = cg.box_cfg}
+    cg.xlog = server:new{alias = 'xlog', box_cfg = cg.box_cfg}
+
+    cg.snap:start()
+    cg.snap:exec(function()
+        t.assert_equals(box.info.id, 1, 'Server id is correct')
+        box.space._schema:replace{'aaa'}
+        t.assert_equals(box.info.lsn, 2, 'Exactly two entries are written')
+        box.snapshot()
+    end)
+    cg.snap:stop()
+    cg.snap_path = fio.pathjoin(cg.snap.workdir, '00000000000000000002.snap')
+    t.assert_equals(#fio.glob(cg.snap_path), 1, 'Snapshot is created')
+
+    cg.xlog:start()
+    cg.xlog:exec(function()
+        t.assert_equals(box.info.id, 1, 'Server id is correct')
+        box.space._schema:replace{'aaa'}
+        box.space._schema:replace{'bbb'}
+        t.assert_equals(box.info.lsn, 3, 'Exactly three entries are written')
+    end)
+    cg.xlog:stop()
+    cg.xlog_path = fio.pathjoin(cg.xlog.workdir, '00000000000000000000.xlog')
+    t.assert_equals(#fio.glob(cg.xlog_path), 1, 'Xlog is created')
+
+    cg.tempdir = fio.tempdir()
+    fio.copyfile(cg.snap_path, cg.tempdir)
+    fio.copyfile(cg.xlog_path, cg.tempdir)
+end)
+
+g.after_all(function(cg)
```

Comment:
Just to be sure (didn't find it in luatest doc): after_all does run always, even if one or more tests fail?

Reply:
Yep, it does. I checked like this:
<details><summary>Applied the difff:</summary>

```diff
diff --git a/test/box-luatest/gh_6794_recover_nonmatching_xlogs_test.lua b/test/box-luatest/gh_6794_>
index 156d8264f..c71830593 100644
--- a/test/box-luatest/gh_6794_recover_nonmatching_xlogs_test.lua
+++ b/test/box-luatest/gh_6794_recover_nonmatching_xlogs_test.lua
@@ -47,6 +47,8 @@ g.after_all(function(cg)
     cg.snap:drop()
     cg.xlog:drop()
     fio.rmtree(cg.tempdir)
+    require("log").info("Run after_all")
+    t.assert(false, 'in after all')
 end)
 
 g.before_test('test_panic_without_force_recovery', function()
@@ -78,6 +80,7 @@ g.test_panic_without_force_recovery = function()
         local filename = fio.pathjoin(g.server.workdir, g.server.alias..'.log')
         t.assert(g.server:grep_log(msg, nil, {filename = filename}))
     end)
+    t.assert(false, 'in test')
 end
 
 g.test_ignore_with_force_recovery = function()
@@ -93,4 +96,5 @@ g.test_ignore_with_force_recovery = function()
         t.assert_equals(box.info.status, 'running', 'The server is recovered')
         t.assert(not box.info.ro, 'Server is writable')
     end)
+    t.assert(false, 'in test')
 end
```
</details>

And ran the test:
```
[001] not ok 1	box-luatest.gh_6794_recover_nonmatching_xlogs.test_panic_without_force_recovery
[001] #   ...t/box-luatest/gh_6794_recover_nonmatching_xlogs_test.lua:83: in test
[001] #   expected: a value evaluating to true, actual: false
[001] #   stack traceback:
[001] #   	...t/box-luatest/gh_6794_recover_nonmatching_xlogs_test.lua:83: in function 'box-luatest.gh_6794_recover_nonmatching_xlogs.test_panic_without_force_recovery'
[001] #   	...
[001] #   	[C]: in function 'xpcall'
[001] #   artifacts:
[001] #   	snapshot -> /tmp/t/001_box-luatest/artifacts/snapshot-7-lkPQAAzFL4
[001] #   	xlog -> /tmp/t/001_box-luatest/artifacts/xlog-xrh0wMBmKACo
[001] #   	master-test_panic -> /tmp/t/001_box-luatest/artifacts/master-test_panic-yXBpjv5Jxbyv
[001] not ok 2	box-luatest.gh_6794_recover_nonmatching_xlogs.test_ignore_with_force_recovery
[001] #   ...t/box-luatest/gh_6794_recover_nonmatching_xlogs_test.lua:99: in test
[001] #   expected: a value evaluating to true, actual: false
[001] #   stack traceback:
[001] #   	...t/box-luatest/gh_6794_recover_nonmatching_xlogs_test.lua:99: in function 'box-luatest.gh_6794_recover_nonmatching_xlogs.test_ignore_with_force_recovery'
[001] #   	...
[001] #   	[C]: in function 'xpcall'
[001] #   artifacts:
[001] #   	master-test_ignore -> /tmp/t/001_box-luatest/artifacts/master-test_ignore-KKQL3-XGVb1n
[001] # Ran 2 tests in 0.771 seconds, 0 succeeded, 2 failed
[001] 
[001] [test-run server "luatest_server"] Last 15 lines of the log file /tmp/t/001_box-luatest/gh_6794_recover_nonmatching_xlogs_test.log:
[001] Run after_all
```
The test logs contain `Run after_all` at `/tmp/t/001_box-luatest/gh_6794_recover_nonmatching_xlogs_test.log`